tracing-subscriber/src/filter/directive.rs

```diff
@@ -455,17 +476,40 @@ impl Statics {
 
 impl PartialOrd for StaticDirective {
     fn partial_cmp(&self, other: &StaticDirective) -> Option<Ordering> {
-        match (self.field_names.len(), other.field_names.len()) {
-            (a, b) if a == b => {}
-            (a, b) => return Some(a.cmp(&b)),
-        }
+        // We attempt to order directives by how "specific" they are. This
+        // ensures that we try the most specific directives first when
+        // attempting to match a piece of metadata.
+
+        // First, we compare based on whether a target is specified, and the
+        // lengths of those targets if both have targets.
+        let ordering = self
+            .target
+            .as_ref()
+            .map(String::len)
+            .cmp(&other.target.as_ref().map(String::len))
+            // Then we compare how many field names are matched by each directive.
+            .then_with(|| self.field_names.len().cmp(&other.field_names.len()))
+            // Finally, we fall back to lexicographical ordering if the directives are
+            // equally specific. Although this is no longer semantically important,
+            // we need to define a total ordering to determine the directive's place
+            // in the BTreeMap.
+            .then_with(|| {
+                self.target.cmp(&other.target).then_with(|| {
+                    FieldOrdering(&self.field_names).cmp(&FieldOrdering(&other.field_names))
```

Comment:
Is the newtype here accomplishing anything, given the explicit precedence given to `field_names.len()` above?

Reply:
I *thought* the behavior of `impl Ord for [T]` should be equivalent, but after removing the newtype, one of the other filtering tests (not the ones that reproduce the bug!) consistently fails. Trying to figure out what causes that.

Reply:
Control only reaches this point if `field_names.len()` is equal, right?

Reply:
Ah, never mind me --- the test was only failing because I had changed something else, I guess. It was late.